src/compiler/moduleNameResolver.ts

```diff
@@ -679,6 +679,17 @@ namespace ts {
             return tryLoadModuleUsingBaseUrl(extensions, moduleName, loader, state);
         }
         else {
+            const { baseUrl, paths } = state.compilerOptions;
+            if (baseUrl && paths && startsWithSlash(moduleName) && getOwnKeys(paths).find(startsWithSlash)) {

```

Comment:
I think we should move this condition into isExternalModuleName and just use the path matching result. The fallback to matching these as relative module name if this condition is satisfied seems sketchy. This would also make all the places that handle the relative/non relative module names work correctly including the resolution caches.

Reply:
How do you propose we do that?

When I look at the code for `tryLoadModuleUsingBaseUrl` I see the exact same mechanism. We first try to resolve it using the alias, if that fails we fall back on other behavior. Here we are doing the same.

This is the code in `isExternalModuleNameRelative`:
```js
pathIsRelative(moduleName) || isRootedDiskPath(moduleName)
```
If we wanted to move the condition here we would need to adjust its signature. We can not just check for the `/`. We actually need to first match against a `paths` pattern and if that fails we need to load it as a 'relative' file.

If I don't explain it correctly please look at the test cases I added.

I you however have a concrete suggestion I am very open to hearing it.

Reply:
I have moved the code into `tryLoadModuleUsingRootDirs` to mirror `tryLoadModuleUsingBaseUrl`. 

Thinking about it, we might want to make it a top-level check. Something like:

```js
if (baseUrl && paths && startsWithSlash(moduleName) && getOwnKeys(paths).find(startsWithSlash)) {
    if (state.traceEnabled) {
        trace(state.host, Diagnostics.baseUrl_option_is_set_to_0_using_this_value_to_resolve_non_relative_module_name_1, baseUrl, moduleName);
        trace(state.host, Diagnostics.paths_option_is_specified_looking_for_a_pattern_to_match_module_name_0, moduleName);
    }
    const resolved = tryLoadModuleUsingPaths(extensions, moduleName, baseUrl, paths, loader, /*onlyRecordFailures*/ false, state);
    if (resolved) {
        return resolved.value;
    }
}

if (!isExternalModuleNameRelative(moduleName)) {
    return tryLoadModuleUsingBaseUrl(extensions, moduleName, loader, state);
}
else {
    return tryLoadModuleUsingRootDirs(extensions, moduleName, containingDirectory, loader, state);
}
```

This however would be a backwards incompatible change with regards to the traces (not sure if that is a problem).

Reply:
I just realized that moving the piece of code out of `tryLoadModuleUsingBaseUrl` and moving the added code out of the if would be potentially a performance hit because there are usually far more `./` imports which should not resolved using path mappings.